### ISSUE TITLE
v8 Deleting a node results in an error

### ISSUE THREAD

**Issue by nul800sebastiaan** [umbraco/Umbraco-CMS#4404]
Trying to delete a node in v8 results in an error, but the node is deleted anyway:

![deletenode](https://user-images.githubusercontent.com/304656/52215951-7cc47a00-2895-11e9-8459-edb9e45fe809.gif)

The error:

```
2019-02-04 15:54:36,435 [P30660/D6/T72] ERROR  Umbraco.Web.Editors.ContentController - Unhandled controller exception occurred for request 'http://localhost:8000/umbraco/backoffice/UmbracoApi/Content/DeleteById?id=1068'
System.Data.SqlServerCe.SqlCeException (0x80004005): A duplicate value cannot be inserted into a unique index. [ Table name = umbracoRelation,Constraint name = IX_umbracoRelation_parentChildType ]
   at System.Data.SqlServerCe.SqlCeCommand.ProcessResults(Int32 hr)
   at System.Data.SqlServerCe.SqlCeCommand.ExecuteCommandText(IntPtr& pCursor, Boolean& isBaseTableCursor)
   at System.Data.SqlServerCe.SqlCeCommand.ExecuteCommand(CommandBehavior behavior, String method, ResultSetOptions options)
   at System.Data.SqlServerCe.SqlCeCommand.ExecuteNonQuery()
   at StackExchange.Profiling.Data.ProfiledDbCommand.ExecuteNonQuery() in c:\TeamCity\buildAgent\work\a743336045964cbc\StackExchange.Profiling\Data\ProfiledDbCommand.cs:line 264
   at Umbraco.Core.Persistence.FaultHandling.FaultHandlingDbCommand.<ExecuteNonQuery>b__31_0() in D:\Dev\Umbraco-CMS8\src\Umbraco.Core\Persistence\FaultHandling\RetryDbConnection.cs:line 209
   at Umbraco.Core.Persistence.FaultHandling.FaultHandlingDbCommand.<>c__DisplayClass33_0`1.<Execute>b__0() in D:\Dev\Umbraco-CMS8\src\Umbraco.Core\Persistence\FaultHandling\RetryDbConnection.cs:line 222
   at Umbraco.Core.Persistence.FaultHandling.RetryPolicy.ExecuteAction[TResult](Func`1 func) in D:\Dev\Umbraco-CMS8\src\Umbraco.Core\Persistence\FaultHandling\RetryPolicy.cs:line 172
   at Umbraco.Core.Persistence.FaultHandling.FaultHandlingDbCommand.Execute[T](Func`1 f) in D:\Dev\Umbraco-CMS8\src\Umbraco.Core\Persistence\FaultHandling\RetryDbConnection.cs:line 219
   at Umbraco.Core.Persistence.FaultHandling.FaultHandlingDbCommand.ExecuteNonQuery() in D:\Dev\Umbraco-CMS8\src\Umbraco.Core\Persistence\FaultHandling\RetryDbConnection.cs:line 209
   at NPoco.Database.ExecuteNonQueryHelper(DbCommand cmd)
   at NPoco.DatabaseTypes.SqlServerCEDatabaseType.ExecuteInsert[T](Database db, DbCommand cmd, String primaryKeyName, Boolean useOutputClause, T poco, Object[] args)
   at NPoco.Database.InsertImp[T](PocoData pocoData, String tableName, String primaryKeyName, Boolean autoIncrement, T poco)
   at NPoco.Database.Insert[T](String tableName, String primaryKeyName, Boolean autoIncrement, T poco)
   at NPoco.Database.Insert[T](T poco)
   at Umbraco.Core.Persistence.Repositories.Implement.RelationRepository.PersistNewItem(IRelation entity) in D:\Dev\Umbraco-CMS8\src\Umbraco.Core\Persistence\Repositories\Implement\RelationRepository.cs:line 142
   at Umbraco.Core.Cache.NoCacheRepositoryCachePolicy`2.Create(TEntity entity, Action`1 persistNew) in D:\Dev\Umbraco-CMS8\src\Umbraco.Core\Cache\NoCacheRepositoryCachePolicy.cs:line 38
   at Umbraco.Core.Persistence.Repositories.Implement.RepositoryBase`2.Save(TEntity entity) in D:\Dev\Umbraco-CMS8\src\Umbraco.Core\Persistence\Repositories\Implement\RepositoryBaseOfTIdTEntity.cs:line 165
   at Umbraco.Core.Services.Implement.RelationService.Save(IRelation relation) in D:\Dev\Umbraco-CMS8\src\Umbraco.Core\Services\Implement\RelationService.cs:line 566
   at Umbraco.Core.Components.RelateOnTrashComponent.ContentService_Trashed(IContentService sender, MoveEventArgs`1 e) in D:\Dev\Umbraco-CMS8\src\Umbraco.Core\Components\RelateOnTrashComponent.cs:line 84
   at Umbraco.Core.Events.EventDefinition`2.RaiseEvent() in D:\Dev\Umbraco-CMS8\src\Umbraco.Core\Events\EventDefinition.cs:line 69
   at Umbraco.Core.Events.QueuingEventDispatcher.ScopeExitCompleted() in D:\Dev\Umbraco-CMS8\src\Umbraco.Core\Events\QueuingEventDispatcher.cs:line 25
   at Umbraco.Core.Events.QueuingEventDispatcherBase.ScopeExit(Boolean completed) in D:\Dev\Umbraco-CMS8\src\Umbraco.Core\Events\QueuingEventDispatcherBase.cs:line 337
   at Umbraco.Core.Scoping.Scope.<>c__DisplayClass72_0.<RobustExit>b__1() in D:\Dev\Umbraco-CMS8\src\Umbraco.Core\Scoping\Scope.cs:line 435
   at Umbraco.Core.Scoping.Scope.TryFinally(Int32 index, Action[] actions) in D:\Dev\Umbraco-CMS8\src\Umbraco.Core\Scoping\Scope.cs:line 474
   at Umbraco.Core.Scoping.Scope.TryFinally(Int32 index, Action[] actions) in D:\Dev\Umbraco-CMS8\src\Umbraco.Core\Scoping\Scope.cs:line 478
   at Umbraco.Core.Scoping.Scope.TryFinally(Action[] actions) in D:\Dev\Umbraco-CMS8\src\Umbraco.Core\Scoping\Scope.cs:line 466
   at Umbraco.Core.Scoping.Scope.RobustExit(Boolean completed, Boolean onException) in D:\Dev\Umbraco-CMS8\src\Umbraco.Core\Scoping\Scope.cs:line 422
   at Umbraco.Core.Scoping.Scope.DisposeLastScope() in D:\Dev\Umbraco-CMS8\src\Umbraco.Core\Scoping\Scope.cs:line 402
   at Umbraco.Core.Scoping.Scope.Dispose() in D:\Dev\Umbraco-CMS8\src\Umbraco.Core\Scoping\Scope.cs:line 365
   at Umbraco.Core.Services.Implement.ContentService.MoveToRecycleBin(IContent content, Int32 userId) in D:\Dev\Umbraco-CMS8\src\Umbraco.Core\Services\Implement\ContentService.cs:line 1677
   at Umbraco.Web.Editors.ContentController.DeleteById(Int32 id) in D:\Dev\Umbraco-CMS8\src\Umbraco.Web\Editors\ContentController.cs:line 1370
   at lambda_method(Closure , Object , Object[] )
   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ActionExecutor.<>c__DisplayClass6_1.<GetExecutor>b__3(Object instance, Object[] methodParameters)
   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ActionExecutor.Execute(Object instance, Object[] arguments)
   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ExecuteAsync(HttpControllerContext controllerContext, IDictionary`2 arguments, CancellationToken cancellationToken)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Controllers.ApiControllerActionInvoker.<InvokeActionAsyncCore>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.ActionFilterAttribute.<ExecuteActionFilterAsyncCore>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.ActionFilterAttribute.<ExecuteActionFilterAsyncCore>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.ActionFilterAttribute.<ExecuteActionFilterAsyncCore>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.ActionFilterAttribute.<ExecuteActionFilterAsyncCore>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.ActionFilterAttribute.<ExecuteActionFilterAsyncCore>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.ActionFilterAttribute.<ExecuteActionFilterAsyncCore>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.ActionFilterAttribute.<ExecuteActionFilterAsyncCore>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Controllers.ActionFilterResult.<ExecuteAsync>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.AuthorizationFilterAttribute.<ExecuteAuthorizationFilterAsyncCore>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.AuthorizationFilterAttribute.<ExecuteAuthorizationFilterAsyncCore>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.AuthorizationFilterAttribute.<ExecuteAuthorizationFilterAsyncCore>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.AuthorizationFilterAttribute.<ExecuteAuthorizationFilterAsyncCore>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Controllers.ExceptionFilterResult.<ExecuteAsync>d__6.MoveNext()

```

**Comment by kjac** [comment]
I have a sneaking suspicion that this is a side effect of #4403 ... chances are if #4403 is solved, this duplicate insert error wouldn't happen.

**Comment by warrenbuckley** [comment]
Closing this as well @nul800sebastiaan as again I can't reproduce this like - https://github.com/umbraco/Umbraco-CMS/issues/4403

**Comment by kjac** [comment]
@warrenbuckley @nul800sebastiaan sorry guys, I can still reproduce this and #4403 on the latest `temp8` using a fresh install (SQL CE) - here's how:

1. Create a root node with one child.
2. Delete the root node.
3. Restore the root node.
4. Kaboom 💥 
5. Reload the browser - the root node is restored, but its relations are all messed up, which now causes #4403 when you attempt to delete it again.

The error does *not* occur if the root has no children. 

The root and child content type can be the same and it doesn't have to be variant content types either.

Here's how it looks:

![delete-error](https://user-images.githubusercontent.com/7405322/52293842-21b78380-2978-11e9-971a-51cd328cccbf.gif)

**Comment by runegronkjaer** [comment]
How come this was closed? I have this problem on my v8.6.2 on an Umbraco Cloud solution.

* It's on my local version of the site.
* It happened after I created some child content on my local and then restored all content from Dev
* The ones I cannot delete are the ones created on my local version and that have never been on Dev or Live

/Rune

**Comment by schrepel** [comment]
I'm experiencing this in 8.6.3 also. Create a content item, save, then try and delete and I get that error. The node does delete despite the error dialog.
